sup3r/utilities/log_interpolator.py

```diff
@@ -0,0 +1,679 @@
+
+        else:
+            with ThreadPoolExecutor(max_workers=max_workers) as exe:
+                for i, file in enumerate(infiles):
+                    outfile = os.path.basename(file).replace('.nc',
+                                                             '_all_interp.nc')
+                    outfile = os.path.join(out_dir, outfile)
+                    futures.append(exe.submit(
+                        cls.run, file, outfile, output_heights=output_heights,
+                        variables=variables,
+                        max_log_height=max_log_height,
+                        overwrite=overwrite))
+                    logger.info(
+                        f'{i + 1} of {len(infiles)} futures submitted.')
+            for i, future in enumerate(as_completed(futures)):
+                future.result()
+                logger.info(f'{i + 1} of {len(futures)} futures complete.')
+
+    @classmethod
+    def pbl_interp_to_height(cls, lev_array, var_array, levels,
+                             fixed_level_mask=None,
+                             max_log_height=100):
+        """Fit ws log law to data below max_log_height.
+
+        Parameters
+        ----------
+        lev_array : ndarray
+            1D Array of height values corresponding to the wrf source
+            data in the same shape as var_array.
+        var_array : ndarray
+            1D Array of variable data, for example u-wind in a 1D array of
+            shape
+        levels : float | list
+            level or levels to interpolate to (e.g. final desired hub heights
+            above surface elevation)
+        fixed_level_mask : ndarray | None
+            Optional mask to use only fixed levels. Fixed levels are those that
+            were not computed from pressure levels but instead added along with
+            wind components at explicit heights (e.g u_10m, v_10m, u_100m,
+            v_100m)
+        max_log_height : int
+            Max height for using log interpolation.
+
+        Returns
+        -------
+        values : ndarray
+            Array of interpolated windspeed values below max_log_height.
+        good : bool
+            Check if log interpolation went without issue.
+        """
+        def ws_log_profile(z, a, b):
+            return a * np.log(z) + b
+
+        lev_array_samp = lev_array.copy()
+        var_array_samp = var_array.copy()
+        if fixed_level_mask is not None:
+            lev_array_samp = lev_array_samp[fixed_level_mask]
+            var_array_samp = var_array_samp[fixed_level_mask]
+
+        good = True
+        levels = np.array(levels)
+        lev_mask = (0 < levels) & (levels <= max_log_height)
+        var_mask = (0 < lev_array_samp) & (lev_array_samp <= max_log_height)
+
+        try:
+            popt, _ = curve_fit(ws_log_profile, lev_array_samp[var_mask],
+                                var_array_samp[var_mask])
+            log_ws = ws_log_profile(levels[lev_mask], *popt)
```

Comment:
well this is nice and clean! all of the optimize methods from scipy are so overwhelming. glad you found one that works.